src/coreclr/jit/runtimelookup.cpp

```diff
@@ -0,0 +1,434 @@
+        cns = (void*)comp->vnStore->CoercedConstantValue<ssize_t>(tree->gtVNPair.GetLiberal());
+    }
+    return cns;
+}
+
+// Save expression to a local and append it as the last statement in exprBlock
+static GenTree* SpillExpression(Compiler* comp, GenTree* expr, BasicBlock* exprBlock, DebugInfo& debugInfo)
+{
+    unsigned const tmpNum         = comp->lvaGrabTemp(true DEBUGARG("spilling expr"));
+    comp->lvaTable[tmpNum].lvType = expr->TypeGet();
+    Statement* asgStmt            = comp->fgNewStmtAtEnd(exprBlock, comp->gtNewTempAssign(tmpNum, expr));
+    asgStmt->SetDebugInfo(debugInfo);
+    comp->gtSetStmtInfo(asgStmt);
+    comp->fgSetStmtSeq(asgStmt);
+    return comp->gtNewLclvNode(tmpNum, expr->TypeGet());
+};
+
+// Create block from the given tree
+static BasicBlock* CreateBlockFromTree(
+    Compiler* comp, BasicBlock* insertAfter, BBjumpKinds blockKind, GenTree* tree, DebugInfo& debugInfo)
+{
+    // Fast-path basic block
+    BasicBlock* newBlock = comp->fgNewBBafter(blockKind, insertAfter, true);
+    newBlock->bbFlags |= BBF_INTERNAL;
+    comp->gtSetEvalOrder(tree);
+    Statement* stmt = comp->fgNewStmtFromTree(tree);
+    comp->fgInsertStmtAtEnd(newBlock, stmt);
+    stmt->SetDebugInfo(debugInfo);
+    comp->gtSetStmtInfo(stmt);
+    comp->fgSetStmtSeq(stmt);
+    comp->gtUpdateStmtSideEffects(stmt);
+    newBlock->bbCodeOffs    = insertAfter->bbCodeOffsEnd;
+    newBlock->bbCodeOffsEnd = insertAfter->bbCodeOffsEnd;
+    return newBlock;
+}
+
+//------------------------------------------------------------------------------
+// fgExpandRuntimeLookups : partially expand runtime lookups helper calls
+//                          to add a nullcheck [+ size check] and a fast path
+// Returns:
+//    PhaseStatus indicating what, if anything, was changed.
+//
+// Notes:
+//    The runtime lookup itself is needed to access a handle in code shared between
+//    generic instantiations. The lookup depends on the typeContext which is only available at
+//    runtime, and not at compile - time. See ASCII block diagrams in comments below for
+//    better understanding how this phase expands runtime lookups.
+//
+PhaseStatus Compiler::fgExpandRuntimeLookups()
+{
+    PhaseStatus result = PhaseStatus::MODIFIED_NOTHING;
+
+    if (!doesMethodHaveExpRuntimeLookup())
+    {
+        // The method being compiled doesn't have expandable runtime lookups. If it does
+        // and doesMethodHaveExpRuntimeLookup() still returns false we'll assert in LowerCall
+        return result;
+    }
+
+    INDEBUG(bool irIsPrinted = false);
+
+    // Find all calls with GTF_CALL_M_EXP_RUNTIME_LOOKUP flag
+    for (BasicBlock* block : Blocks())
+    {
+        for (Statement* const stmt : block->Statements())
+        {
+            if ((stmt->GetRootNode()->gtFlags & GTF_CALL) == 0)
+            {
+                // TP: Stmt has no calls - bail out
+                continue;
+            }
+
+            for (GenTree* const tree : stmt->TreeList())
```

Comment:
this is exactly why I had here "do re-scan current block again" - I think it's better to do that. The current statement will likely (always) be the first statement of the current block after split

Reply:
added

Reply:
Ah ok, I thought restarting in the block was starting over on statements we already processed again.

Reply:
We modify `block` in the loop but the `block` used internally in the iterator for the loop is something else. It makes me quite uneasy -- maybe it's better to write this without range-based `for`, using some `BasicBlock* nextBlock` instead?

Reply:
e.g. something like the following would make me feel a bit more at ease:
```cpp
for (BasicBlock* block = FirstBlock(); block != nullptr;)
{
  BasicBlock* nextBlock = block->NextBlock();
  for (Statement* stmt : block->Statements())
  {
    if (ExpandOneRuntimeLookup(stmt, &nextBlock))
    {
      break;
    }
  }

  block = nextBlock;
}
```

Reply:
I'm not sure I understand how we possible can get a different block after `block`? when we split we basically add new blocks in front of it and the currently iterated block will remain unchanged

Reply:
> when we split we basically add new blocks in front of it

Is that true? If I read `fgSplitBlockAfterStatement` which uses `fgSplitBlockAtEnd` correctly then that's not true -- it adds the new block after the current block and returns the new block. The code then does  `block = <the after block>` and does the goto. This will visit all statements in `block` again, but when we get back to the iterator at some point it still contains the original "before" block, and the next block after that is something we don't need to revisit (including `<the after block>` at some point).

Reply:
Yes, you're right. I've just checked. So nothing bad happens but the next block after `block` via iteration will be `nullcheckBb` we created - so it will traverse internal blocks we created previously. Let me change that to normal for loop like you suggested

Reply:
@jakobbotsch changed and verified that block iteration works as expected - it re-visit current block again (which will be the last one in the shape we create) and the next one will be the actual bbNext